### DISCUSSION TITLE
vectorStore using TypeORM experiencing connection error on Chain retriever

### DISCUSSION THREAD

**Discussion by fabiovalencio** [FlowiseAI/Flowise#738] [Q&A] [answered]
Hi there, 

I am trying to create a vector stores component using TypeORM: https://js.langchain.com/docs/modules/data_connection/vectorstores/integrations/typeorm.

`
...

     init(nodeData, _, options) {

        const credentialData = await (0, utils_1.getCredentialData)(nodeData.credential ?? '', options);
        const user = (0, utils_1.getCredentialParam)('user', credentialData, nodeData);
        const password = (0, utils_1.getCredentialParam)('password', credentialData, nodeData);
        const tableName = nodeData.inputs?.tableName;
        const docs = nodeData.inputs?.document;
        const embeddings = nodeData.inputs?.embeddings;
        const output = nodeData.outputs?.output;
        const topK = nodeData.inputs?.topK;
        const k = topK ? parseFloat(topK) : 4;
        const postgresConnectionOptions = {
            type: "postgres",
            host: nodeData.inputs?.host,
            port: nodeData.inputs?.port,
            username: user,
            password: password,
            database: nodeData.inputs?.database
        }

        const args = {
            postgresConnectionOptions: postgresConnectionOptions,
            tableName: tableName,
        }
        const flattenDocs = docs && docs.length ? (0, lodash_1.flatten)(docs) : [];
        const finalDocs = [];
        for (let i = 0; i < flattenDocs.length; i += 1) {
            finalDocs.push(new document_1.Document(flattenDocs[i]));
        }
        const vectorStore = await typeorm_1.TypeORMVectorStore.fromDocuments(finalDocs, embeddings, args);
        if (output === 'retriever') {
            const retriever = vectorStore.asRetriever(k);
            return retriever;
        } else if (output === 'vectorStore') {
            vectorStore.k = k;
            return vectorStore;
        }
        return vectorStore;
    }
`

So far so good, the vectors are saved in the database and the component returns the retriever.
But the chain (conversational retriever qa chain) seems to be unable to connect to the database and breaks the application.
`
[chain/start] [1:chain:ConversationalRetrievalQAChain] Entering Chain run with input: {
  "question": "what the document is about",
  "chat_history": []
}
`
`
[retriever/start] [1:chain:ConversationalRetrievalQAChain > 2:retriever:VectorStoreRetriever] Entering Retriever run with input: {
  "query": "what the document is about"
}
`
`
2023-08-11 10:36:33 [ERROR]: uncaughtException:  Illegal invocation
TypeError: Illegal invocation
    at Socket.ref (node:net:1524:18)
    at Connection.ref (.../node_modules/pg/lib/connection.js:183:17)
    at Client.ref (.../node_modules/pg/lib/client.js:591:21)
    at BoundPool._pulseQueue (/node_modules/pg-pool/index.js:148:28)
    at .../node_modules/pg-pool/index.js:184:37
    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)
`

I have been struggling with this for couple of days but to no avail.
I would like to make available to the community a generic vector store for the PostgreSQL database.

![Screenshot 2023-08-11 at 11 07 40](https://github.com/FlowiseAI/Flowise/assets/215687/e593cce9-5163-4402-9826-3efdc72c973f)

Can someone please guide me? Thanks.

**Comment by chungyau97** [comment, marked as answer]
Hi @fabiovalencio,

Could you submit a PR so that I can look into it?
